BootloaderCommonPkg/Library/PagingLib/PagingMap.c

```diff
@@ -44,60 +50,100 @@ MapMemoryRange (
+  // Check for Mapping PageSize alignment
+  if ( (Ranges[0].Mapping & (Ranges[0].PageSize - 1)) != 0) {
     return EFI_INVALID_PARAMETER;
   }
 
-  // Create 4KB pages and remap the CAR region into a different memory location
-  PageTable4K = (UINTN *)((UINTN)PageBuffer + PteOffset);
-  Address = PageBase;
-  for (Idx = 0; Idx < EntryNum; Idx++) {
-    if ((Address >= Ranges[0].Start) && (Address <= Ranges[0].Limit)) {
-      PageTable4K[Idx]   = Ranges[0].Mapping + (Attribute | IA32_PG_PD);
-      Ranges[0].Mapping += SIZE_4KB;
+  switch (Ranges[0].PageSize) {
+  case SIZE_4KB:
+    PageBits  = GetPageShiftBits (PDE_PG_LVL);
+    if (PageBits == 21) {
+      // 64 bit mode
+      PteOffset = 6 * EFI_PAGE_SIZE;
+      PdeOffset = 2 * EFI_PAGE_SIZE;
+      EntryNum  = 512;
     } else {
-      PageTable4K[Idx]   = Address + (Attribute | IA32_PG_PD);
+      // 32 bit mode
+      PteOffset = 1 * EFI_PAGE_SIZE;
+      PdeOffset = 0;
+      EntryNum  = 1024;
     }
-    Address += SIZE_4KB;
+    PageTable4K = (UINTN *)((UINTN)PageBuffer + PteOffset);
+    PageTable2M = (UINTN *)((UINTN)PageBuffer + PdeOffset);
+    break;
+  case SIZE_1GB:
+    ASSERT (Ranges[0].Limit < SIZE_512GB);
+    PageBits    = GetPageShiftBits (PDP_PG_LVL);
+    PdpOffset   = 1 * EFI_PAGE_SIZE;
+    EntryNum    = 512;
+    PageTable1G = (UINTN *)((UINTN)PageBuffer + PdpOffset);
+    break;
+  default:
+    return EFI_INVALID_PARAMETER;
   }
 
-  // Split the 2MB page containing the CAR region into 4KB pages
-  Idx = (UINT32) RShiftU64 (PageBase, PageBits);
-  PageTable      = (UINTN *)((UINTN)PageBuffer + PdeOffset);
-  PageTable[Idx] = (UINTN)PageTable4K + Attribute;
+  Alignment = (UINTN) LShiftU64 (1, PageBits) - 1;
+  PageBase  = Ranges[0].Start & ~Alignment;
+  Address   = PageBase;
+
+  switch (Ranges[0].PageSize) {
+  case SIZE_4KB:
+    // Check if the limit is in the same page
+    if ((Ranges[0].Limit & ~Alignment) != PageBase) {
+      return EFI_INVALID_PARAMETER;
+    }
+
+    // Create 4KB pages and remap the CAR region into a different memory location
+    for (Idx = 0; Idx < EntryNum; Idx++) {
+      if ((Address >= Ranges[0].Start) && (Address <= Ranges[0].Limit)) {
+        PageTable4K[Idx]   = Ranges[0].Mapping + (Attribute | IA32_PG_PD);
+        Ranges[0].Mapping += SIZE_4KB;
+      } else {
+        PageTable4K[Idx]   = Address + (Attribute | IA32_PG_PD);
+      }
+      Address += SIZE_4KB;
+    }
+
+    // Split the 2MB page containing the CAR region into 4KB pages
+    Idx = (UINT32) RShiftU64 (PageBase, PageBits);
+    PageTable2M[Idx] = (UINTN)PageTable4K + Attribute;
+
+    break;
+  case SIZE_1GB:
+    // Get the start and ending index into 1G table and add 1G entries
+    Idx = (UINT32) RShiftU64 (PageBase, PageBits);
+    End = (UINT32) RShiftU64 (Ranges[0].Limit + 1, PageBits);
+    for (; Idx < End; Idx++) {
+      PageTable1G[Idx] = Address + (Attribute | IA32_PG_PD);

```

Comment:
Address should be set to start from "Ranges[0].Mapping" above line #136.

Reply:
What happens for a 1:1 mapping? Should the caller set Mapping = Start before passing Ranges param into this function? What about lines #108, 109?

Reply:
yes, for 1:1, the caller needs to make them identical.
#109 should only be used for 4Kb page case.